libpoolprotocols/stratum/EthStratumClient.cpp

```diff
@@ -1511,45 +1522,56 @@ void EthStratumClient::sendSocketData(Json::Value const& jReq)
     if (!isConnected())
         return;
 
-    // Out received message only for debug purpouses
-    if (g_logOptions & LOG_JSON)
-    {
-        cnote << jReq;
-    }
-
-    std::ostream os(&m_sendBuffer);
-    os << m_jWriter.write(jReq);  // Do not add lf. It's added by writer.
-
-    if (m_conn->SecLevel() != SecureLevel::NONE)
-    {
-        async_write(*m_securesocket, m_sendBuffer,
-            m_io_strand.wrap(boost::bind(&EthStratumClient::onSendSocketDataCompleted, this,
-                boost::asio::placeholders::error)));
-    }
-    else
-    {
-        async_write(*m_nonsecuresocket, m_sendBuffer,
-            m_io_strand.wrap(boost::bind(&EthStratumClient::onSendSocketDataCompleted, this,
-                boost::asio::placeholders::error)));
-    }
+    // Need a copy of the value for the Q, parameter is ephemerous
+    Json::Value *j = new Json::Value(jReq);
+    m_sendQ.push(j);
+    // signal the thread
+    m_sendQueued = true;
 }
 
-void EthStratumClient::onSendSocketDataCompleted(const boost::system::error_code& ec)
+void EthStratumClient::processSendQ()
 {
-    if (ec)
+    while (!m_sendExit)
     {
-        if ((ec.category() == boost::asio::error::get_ssl_category()) &&
-            (SSL_R_PROTOCOL_IS_SHUTDOWN == ERR_GET_REASON(ec.value())))
-        {
-            cnote << "SSL Stream error :" << ec.message();
-            m_io_service.post(m_io_strand.wrap(boost::bind(&EthStratumClient::disconnect, this)));
-        }
+        m_sendQueued.wait(true);
+        m_sendQueued = false;
 
-        if (isConnected())
+        Json::Value* j;
+        while (m_sendQ.pop(j))
         {
-            dev::setThreadName("stratum");
-            cwarn << "Socket write failed: " + ec.message();
-            m_io_service.post(m_io_strand.wrap(boost::bind(&EthStratumClient::disconnect, this)));
+            // Out received message only for debug purpouses
+            if (g_logOptions & LOG_JSON)
+            {
+                cnote << *j;
+            }
+
+            std::ostream os(&m_sendBuffer);
+            os << m_jWriter.write(*j);  // Do not add lf. It's added by writer.
+            delete j;
+
+            boost::system::error_code ec;
+            if (m_conn->SecLevel() != SecureLevel::NONE)
+                write(*m_securesocket, m_sendBuffer, ec);
+            else
+                write(*m_nonsecuresocket, m_sendBuffer, ec);
+            if (ec)
+            {
+                if ((ec.category() == boost::asio::error::get_ssl_category()) &&
```

Comment:
If you get here you should exit the inner loop and purge the sendQ as, in case you're processing two or messages, you'd get two or more errors and would issue two or more disconnects